openmdao/components/interp_util/interp_bsplines.py

```diff
@@ -0,0 +1,201 @@
+
+    def initialize(self):
+        """
+        Declare options.
+        """
+        self.options.declare('order', default=4,
+                             desc='B-spline order.')
+
+    def check_config(self):
+        """
+        Verify that we have enough points for this interpolation algorithm.
+        """
+        pass
+
+    def evaluate_vectorized(self, x):
+        """
+        Interpolate across all table dimensions for all requested samples.
+
+        Parameters
+        ----------
+        x : ndarray
+            The coordinates to sample the gridded data at. First array element is the point to
+            interpolate here. Remaining elements are interpolated on sub tables.
+
+        Returns
+        -------
+        ndarray
+            Interpolated values.
+        ndarray
+            Derivative of interpolated values with respect to this independents.
+        ndarray
+            Derivative of interpolated values with respect to values.
+        ndarray
+            Derivative of interpolated values with respect to grid.
+        """
+        n_cp = self.values.shape[-1]
+        if self._jac is None:
+            self._jac = self.get_bspline_mtx(n_cp, x / x[-1],
+                                             order=self.options['order']).tocoo()
+
+        result = np.einsum('ij,kj->ki', self._jac.toarray(), self.values)
+
+        return result, None, self._jac, None
+
+    def training_gradients(self, pt):
+        """
+        Compute the training gradient for the vector of training points.
+
+        Parameters
+        ----------
+        pt : ndarray
+            Training point values.
+
+        Returns
+        -------
+        ndarray
+            Gradient of output with respect to training point values.
+        """
+        return self._jac.toarray().flatten()
+
+    def get_bspline_mtx(self, num_cp, t_vec, order=4):
+        """
+        Compute matrix of B-spline coefficients.
+
+        Parameters
+        ----------
+        num_cp : int
+            Number of control points.
+        t_vec : int
+            Interpolated point locations.
+        order : int(4)
+            B-spline order.
+
+        Returns
+        -------
+        csr_matrix
+            Sparse matrix of B-spline coefficients.
+        """
+        knots = np.zeros(num_cp + order)
+        knots[order - 1:num_cp + 1] = np.linspace(0, 1, num_cp - order + 2)
+        knots[num_cp + 1:] = 1.0
+
+        basis = np.zeros(order, dtype=t_vec.dtype)
+        arange = np.arange(order)
+
+        num_pt = len(t_vec)
+        data = np.zeros((num_pt, order), dtype=t_vec.dtype)
+        rows = np.zeros((num_pt, order), int)
+        cols = np.zeros((num_pt, order), int)
+
+        for ipt in range(num_pt):
+            t = t_vec[ipt]
+
+            i0 = -1
+            for ind in range(order, num_cp + 1):
+                if (knots[ind - 1].real <= t.real) and (t.real < knots[ind].real):
+                    i0 = ind - order
```

Comment:
looks like we're assigning the value of i0 multiple times in this loop.  Should there be a `break` after the assignment?

Reply:
yeah, given that knots is non-decreasing, we should be able to break here.